FrameworkInternals/meta_build_info.py

```diff
@@ -0,0 +1,55 @@
+import os
+import socket
+import sys
+import argparse
+import version_control_interface
+from argparse import RawTextHelpFormatter
+from time import localtime, strftime
+
+def process_args():
+	print 'meta_build_info.py, processing command line args...'
+	parser = argparse.ArgumentParser(description="""meta_build_info.py generates build info. Requires 2 args - the ABS path to the target generation directory (must already exist) and the toolkit """, formatter_class=RawTextHelpFormatter)
+	parser.add_argument('--target_generation_dir', required=True, help='absolute path to local directory where the build meta info will be generated')
+	parser.add_argument('--toolkit_libs', required=True, help='which opc-ua toolkit libraries was this server built with')
+	args = parser.parse_args()
+	print 'generating to directory [{0}], toolkit libs [{1}]'.format(args.target_generation_dir, args.toolkit_libs)
+	return args
+
+def open_generation_target_file(target_generation_dir):
+	if not os.path.exists(target_generation_dir):
+		sys.exit('Generation Failed! Could not find target generation dir [{0}]'.format(target_generation_dir))
+	generation_file_path = os.path.join(target_generation_dir,  'MetaBuildInfoGenerated.h')
+	try:
+		return open(generation_file_path, 'w')
+	except Exception, e:
+		sys.exit('Generation Failed! Could not open/create target generation file [{0}]'.format(generation_file_path))
+
+def generateBuildTime(generatedFile):
+	now = strftime("%a, %d %b %Y %H:%M:%S", localtime())
+	generatedFile.write('#define GEND_BUILD_TIME "{0}"\n'.format(now))
+
+def generateBuildHost(generatedFile):
+	build_host = socket.gethostname()
+	generatedFile.write('#define GEND_BUILD_HOST "{0}"\n'.format(build_host))
+
+def generateCommitID(generatedFile):
+	commitID = "No commit ID found"
+	try:
+		project_root_path = os.path.dirname(os.path.realpath(__file__))+os.path.sep+'..'
```

Comment:
I'd suggest passing value of PROJECT_SOURCE_DIR from master CMakeLists into here (via command line arg or so).  I'm not sure if this assumption really scales well.

Reply:
Yep, I'll try that. The existing code is a little nasty.

Reply:
Done
Meta/CMakeLists.txt
```
COMMAND python ${CMAKE_CURRENT_LIST_DIR}/../FrameworkInternals/meta_build_info.py --project_root_dir ${CMAKE_SOURCE_DIR} --target_generation_dir ${    PROJECT_BINARY_DIR}/generated --toolkit_libs "${OPCUA_TOOLKIT_LIBS}"
```

Then in meta_build_info.py
```
def process_args():
print 'meta_build_info.py, processing command line args...'
parser = argparse.ArgumentParser(description="""meta_build_info.py generates build info. Requires 3 args - Abs path to project root, abs path to th    e target generation directory (must already exist) and the toolkit libs""", formatter_class=RawTextHelpFormatter)
parser.add_argument('--project_root_dir', required=True, help='absolute path to quasar project root directory')
...

#used here

def generateCommitID(generatedFile, project_root_dir):
commitID = "No commit ID found"
try:
        commitID = version_control_interface.VersionControlInterface(project_root_dir).get_latest_repo_commit()
```